module/zfs/zio_checksum.c

```diff
@@ -545,14 +545,40 @@ zio_checksum_error(zio_t *zio, zio_bad_cksum_t *info)
 	uint_t checksum = (bp == NULL ? zio->io_prop.zp_checksum :
 	    (BP_IS_GANG(bp) ? ZIO_CHECKSUM_GANG_HEADER : BP_GET_CHECKSUM(bp)));
 	int error;
-	uint64_t size = (bp == NULL ? zio->io_size :
-	    (BP_IS_GANG(bp) ? SPA_GANGBLOCKSIZE : BP_GET_PSIZE(bp)));
+	uint64_t size = bp == NULL ? zio->io_size :
+	    (BP_IS_GANG(bp) ? SPA_OLD_GANGBLOCKSIZE : BP_GET_PSIZE(bp));
 	uint64_t offset = zio->io_offset;
 	abd_t *data = zio->io_abd;
 	spa_t *spa = zio->io_spa;
 
+	if (bp && BP_IS_GANG(bp) && spa_feature_is_active(spa,
+	    SPA_FEATURE_DYNAMIC_GANG_HEADER)) {
+		size = UINT64_MAX;
+		for (int dva = 0; dva < BP_GET_NDVAS(bp); dva++) {
+			vdev_t *vd = vdev_lookup_top(spa,
+			    DVA_GET_VDEV(&bp->blk_dva[dva]));
+			size = MIN(size, vdev_gang_header_asize(vd));
+		}
```

Comment:
Do we need this part here?  Shouldn't it already be computed by `zio_gang_tree_assemble()` and available here as `zio->io_size`?